### PR TITLE
[SYCLomatic] Introduce an inline asm parser

### PR DESCRIPTION
**Pull request by yihwang-nv**
Signed-off-by: Wang, Yihan [yihan.wang@intel.com](mailto:yihan.wang@intel.com)

### CODE REVIEW

File: clang/lib/DPCT/AsmMigration.cpp [hunk 1]
```diff
@@ -0,0 +1,687 @@
...
+  MF.addMatcher(
+      asmStmt(hasAncestor(functionDecl(
+                  anyOf(hasAttr(attr::CUDADevice), hasAttr(attr::CUDAGlobal)))))
+          .bind("asm"),
+      this);
+}
+
+void AsmRule::runRule(const ast_matchers::MatchFinder::MatchResult &Result) {
+  if (auto *AS = getNodeAsType<AsmStmt>(Result, "asm")) {
+    if (const auto *Asm = dyn_cast<GCCAsmStmt>(AS)) {
```
Comment (tangjj11) [inline clang/lib/DPCT/AsmMigration.cpp]: Will it be` MSAsmStmt` class type?

File: clang/lib/DPCT/AsmMigration.cpp [hunk 2]
```diff
@@ -0,0 +1,687 @@
...
+      std::string S = Asm->generateAsmString(C);
+      PtxContext Context;
+      std::string Replacement;
+      llvm::raw_string_ostream OS(Replacement);
+      llvm::SourceMgr Mgr;
+      Mgr.AddNewSourceBuffer(
+          llvm::MemoryBuffer::getMemBuffer(Asm->getAsmString()->getString()),
+          llvm::SMLoc());
+      PtxParser Parser(Context, Mgr);
+      SYCLGen CodeGen(OS);
```
Comment (tangjj11) [inline clang/lib/DPCT/AsmMigration.cpp]: It more like `DisAsm` but not `CodeGen`.

File: clang/lib/DPCT/AsmMigration.cpp [hunk 3]
```diff
@@ -0,0 +1,687 @@
...
+      unsigned OperandIdx = 0;
+      std::string AsmString;
+
+      auto getReplaceString = [&](const Expr *E) {
+        ExprAnalysis EA;
+        EA.analyze(E);
+        if (isa<IntegerLiteral>(E) || isa<DeclRefExpr>(E) ||
```
Comment (tangjj11) [inline clang/lib/DPCT/AsmMigration.cpp]: Can use `needExtraParens(E)` instead?

File: clang/lib/DPCT/AsmMigration.cpp [hunk 4]
```diff
@@ -0,0 +1,687 @@
...
+            isa<ImplicitCastExpr>(E)) {
+          return EA.getReplacedString();
+        }
+        return "(" + EA.getReplacedString() + ")";
+      };
+
+      for (const auto *E : Asm->outputs()) {
```
Comment (tangjj11) [inline clang/lib/DPCT/AsmMigration.cpp]: Can use `Asm->children()` to replace these two `for` loop.

File: clang/lib/DPCT/Asm/AsmLexer.cpp [hunk 5]
```diff
@@ -0,0 +1,723 @@
...
+      .Case(".global", true)
+      .Case(".local", true)
+      .Case(".param", true)
+      .Case(".shared", true)
+      .Case(".tex", true)
+      .Default(false);
+}
+
+bool PtxToken::isTypeName() const {
+  static StringSet<> TypeNames{".s2",     ".s4",     ".s8",   ".s16",   ".s32",
```
Comment (tangjj11) [inline clang/lib/DPCT/Asm/AsmLexer.cpp]: Can be `static const`.

File: clang/lib/DPCT/Asm/AsmParser.h [hunk 6]
```diff
@@ -0,0 +1,1063 @@
...
+
+enum BooleanOp { BO_And, BO_Or, BO_Xor };
+
+enum ArithmeticOp { AO_Add, AO_Min, AO_Max, AO_Maxabs, AO_Popc };
+
+InstKind FindInstructionKindFromName(StringRef InstName);
+
+} // namespace ptx
+
+class PtxType;
```
Comment (tangjj11) [inline clang/lib/DPCT/Asm/AsmParser.h]: Do we need a `AsmAST.h` file?

File: clang/lib/DPCT/AsmMigration.cpp (hunk 1 above)
Comment (yihwang-nv) [reply clang/lib/DPCT/AsmMigration.cpp]: Yes, but I have not found any ms-style inline asm code in https://docs.nvidia.com/cuda/inline-ptx-assembly/index.html, and this patch will only support gas style inline asm.

File: clang/lib/DPCT/AsmMigration.cpp (hunk 2 above)
Comment (yihwang-nv) [reply clang/lib/DPCT/AsmMigration.cpp]: I think this stage is more like codegen, It's generated SYCL code from an AST.

File: clang/lib/DPCT/AsmMigration.cpp (hunk 3 above)
Comment (yihwang-nv) [reply clang/lib/DPCT/AsmMigration.cpp]: +1

File: clang/lib/DPCT/AsmMigration.cpp (hunk 4 above)
Comment (yihwang-nv) [reply clang/lib/DPCT/AsmMigration.cpp]: This action add inline asm placeholder into symbol table as builtin/predefine symbols, it's requires output operands at the front of input operands, If `Asm->children()` have the order guarantee, then we can use `Asm->children()`.

File: clang/lib/DPCT/Asm/AsmLexer.cpp (hunk 5 above)
Comment (yihwang-nv) [reply clang/lib/DPCT/Asm/AsmLexer.cpp]: +1

**Review by AndyCHHuang — APPROVED** [review]
LGTM

File: clang/lib/DPCT/Asm/AsmIdentifierTable.h [hunk 7]
```diff
@@ -0,0 +1,192 @@
...
+#define CLANG_DPCT_ASM_IDENTIFIER_TABLE_H
+
+#include "AsmTokenKinds.h"
+#include "clang/Basic/LLVM.h"
+#include "llvm/ADT/StringMap.h"
+#include "llvm/Support/Allocator.h"
+
+namespace clang::dpct {
+
+class DpctAsmIdentifierInfo {
```
Comment (zhimingwang36) [inline clang/lib/DPCT/Asm/AsmIdentifierTable.h]: Dpct prefix is not required for the class name as there are already in dpct:: namespace, let's remove the Dpct prefix.

File: clang/lib/DPCT/Asm/AsmIdentifierTable.h [hunk 8]
```diff
@@ -0,0 +1,192 @@
...
+  /// Return true if this token is a builtin type in asm.
+  bool isBuiltinType() const;
+
+  /// Provide less than operator for lexicographical sorting.
+  bool operator<(const DpctAsmIdentifierInfo &RHS) const {
+    return getName() < RHS.getName();
+  }
+};
+
+class DpctAsmIdentifierInfoLookup {
```
Comment (zhimingwang36) [inline clang/lib/DPCT/Asm/AsmIdentifierTable.h]: same as above,  and please check other Dpct prefix

File: clang/lib/DPCT/Asm/AsmIdentifierTable.h (hunk 7 above)
Comment (zhimingwang36) [inline clang/lib/DPCT/Asm/AsmIdentifierTable.h]: add comment to tell the function of the class.

File: clang/lib/DPCT/Asm/AsmIdentifierTable.h (hunk 8 above)
Comment (zhimingwang36) [inline clang/lib/DPCT/Asm/AsmIdentifierTable.h]: add comment to function of looup class.

File: clang/lib/DPCT/Asm/AsmLexer.cpp [hunk 9]
```diff
@@ -0,0 +1,399 @@
...
+#include "llvm/ADT/SmallString.h"
+#include "llvm/ADT/StringExtras.h"
+#include "llvm/ADT/StringSet.h"
+#include "llvm/Support/MemoryBufferRef.h"
+#include "llvm/Support/SaveAndRestore.h"
+#include "llvm/Support/raw_ostream.h"
+
+using namespace clang::dpct;
+
+DpctAsmLexer::DpctAsmLexer(llvm::MemoryBufferRef Input)
```
Comment (zhimingwang36) [inline clang/lib/DPCT/Asm/AsmLexer.cpp]: Please remove the Dpct prefix.

File: clang/lib/DPCT/Asm/AsmParser.cpp [hunk 10]
```diff
@@ -0,0 +1,1271 @@
...
+    break;
+  default:
+    llvm_unreachable("unexpected op token");
+  }
+
+  return ::new (Context)
+      DpctAsmUnaryOperator(Opcode, SubExpr, SubExpr->getType());
+}
+
+// clang-format off
```
Comment (zhimingwang36) [inline clang/lib/DPCT/Asm/AsmParser.cpp:668]: why set clang-format off for the code block?

File: clang/lib/DPCT/Asm/AsmParser.cpp [hunk 11]
```diff
@@ -0,0 +1,1271 @@
...
+  }
+  return Opc;
+}
+// clang-format on
+
+DpctAsmExprResult DpctAsmParser::ActOnBinaryOp(asmtok::TokenKind OpTok,
+                                               DpctAsmExpr *LHS,
+                                               DpctAsmExpr *RHS) {
+  DpctAsmBinaryOperator::Opcode Opcode = ConvertTokenKindToBinaryOpcode(OpTok);
+  /// TODO: Compute the type of binary operator
```
Comment (zhimingwang36) [inline clang/lib/DPCT/Asm/AsmParser.cpp:702]: more detail on the TODO?

File: clang/test/dpct/asm_lop3.cu [hunk 12]
```diff
@@ -30,7 +30,8 @@ static __device__ __forceinline__ uint32_t LOP3LUT_ANDOR(uint32_t a, uint32_t b)
 #define B 3
 __device__  int hard(int a) {
   int d4;
-  // CHECK: d4 = (~((a + B) * (a + B)) & B & ~(3)) | (~((a + B) * (a + B)) & B & (3)) | (((a + B) * (a + B)) & ~B & ~(3));
-  asm("lop3.b32 %0, %1 * %1, %2, 3, 0x1C;" : "=r"(d4) : "r"(a + B), "r"(B));
+  // CHECK: d4 = (~(a + B) & B & ~3) | (~(a + B) & B & 3) | ((a + B) & ~B & ~3);
```
Comment (zhimingwang36) [inline clang/test/dpct/asm_lop3.cu:33]: why update this test case.

File: clang/lib/DPCT/AsmMigration.cpp [hunk 13]
```diff
@@ -0,0 +1,785 @@
...
+#include <iterator>
+#include <limits>
+#include <sstream>
+
+using namespace clang;
+using namespace clang::dpct;
+
+namespace {
+
+class SYCLGenBase {
```
Comment (zhimingwang36) [inline clang/lib/DPCT/AsmMigration.cpp:51]: add comment for SYCLgenBase.

File: clang/lib/DPCT/AsmMigration.cpp [hunk 14]
```diff
@@ -0,0 +1,785 @@
...
+      for (unsigned I = 0, E = AS->getNumInputs(); I != E; ++I)
+        Parser.addInlineAsmOperands(getReplaceString(AS->getInputExpr(I)),
+                                    AS->getInputConstraint(I));
+
+      CodeGen.setNumIndent(1);
+      CodeGen.setIndentUnit(
+          getIndent(AS->getBeginLoc(), DpctGlobalInfo::getSourceManager()));
+
+      do {
+        auto Inst = Parser.ParseStatement();
```
Comment (zhimingwang36) [inline clang/lib/DPCT/AsmMigration.cpp]: Please add  more test against the  Parser,  like lit.

File: clang/test/dpct/asm_lop3.cu (hunk 12 above)
Comment (yihwang-nv) [reply clang/test/dpct/asm_lop3.cu:33]: This case is ill-formed, and rejected by nvcc, as far as I know, the asm syntax does not allow addition of numeric constants and register variables.

**Comment by keryell** [comment]
Interesting idea! :-)

File: clang/lib/DPCT/Asm/AsmIdentifierTable.h (hunk 7 above)
Comment (yihwang-nv) [reply clang/lib/DPCT/Asm/AsmIdentifierTable.h]: Update DpctAsm -> InlineAsm

File: clang/lib/DPCT/AsmMigration.cpp [hunk 15]
```diff
@@ -0,0 +1,786 @@
...
+#include <limits>
+#include <sstream>
+
+using namespace clang;
+using namespace clang::dpct;
+
+namespace {
+
+class SYCLGenBase {
+  llvm::StringMap<std::string> NameAliasTable;
```
Comment (zhimingwang36) [inline clang/lib/DPCT/AsmMigration.cpp]: NameAliasTable dead code?

File: clang/lib/DPCT/AsmMigration.cpp [hunk 16]
```diff
@@ -0,0 +1,786 @@
...
+  case asmtok::op_##X:                                                         \
+    return handle_##X(I);
+#include "Asm/AsmTokenKinds.def"
+  default:
+    break;
+  }
+  return true;
+}
+
+bool SYCLGenBase::emitGuardInstruction(
```
Comment (zhimingwang36) [inline clang/lib/DPCT/AsmMigration.cpp]:
```suggestion
bool SYCLGenBase::emitConditionInstruction(
```

File: clang/lib/DPCT/AsmMigration.cpp [hunk 17]
```diff
@@ -0,0 +1,786 @@
...
+        break;
+      case asmtok::kw_hs:
+        if (T->isUnsignedInt())
+          Template = "{0} >= {1}";
+        else
+          return true;
+        break;
+      case asmtok::kw_equ:
+        if (T->isFloating())
+          Template = "{0} == {1} || sycl::isnan({0}) || sycl::isnan({1})";
```
Comment (zhimingwang36) [inline clang/lib/DPCT/AsmMigration.cpp]:
```suggestion
          Template = "sycl::isnan({0}) || sycl::isnan({1}) || {0} == {1} ";
```

File: clang/lib/DPCT/Asm/AsmParser.h [hunk 18]
```diff
@@ -0,0 +1,1478 @@
...
+};
+
+/// This represents one declaration (or definition), e.g. a variable,
+/// label, etc.
+class InlineAsmDecl {
+public:
+  enum DeclClass {
+    VariableDeclClass,
+
+    /// FIXME: Label declaration do now support now.
```
Comment (zhimingwang36) [inline clang/lib/DPCT/Asm/AsmParser.h]:
```suggestion
    /// FIXME: Label declaration do not support now.
```

File: clang/lib/DPCT/Asm/AsmParser.h [hunk 19]
```diff
@@ -0,0 +1,1478 @@
...
+  /// The type attributes in this instruction.
+  /// e.g. instruction mov.s32 val, 123; has a type attribute 's32'.
+  SmallVector<InlineAsmType *, 4> Types;
+
+  /// This represents arrtibutes like: comparsion operator, rounding modifiers,
+  /// ... e.g. instruction setp.eq.s32 has a comparsion operator 'eq'.
+  SmallVector<InlineAsmIdentifierInfo *, 4> Attributes;
+  InlineAsmExpr *OutputOperand;
+  SmallVector<InlineAsmExpr *, 4> InputOperands;
+  InlineAsmExpr *PredOutput = nullptr;
```
Comment (zhimingwang36) [inline clang/lib/DPCT/Asm/AsmParser.h:334]: what is predOutput? please give example here

File: clang/lib/DPCT/Asm/AsmParser.h [hunk 20]
```diff
@@ -0,0 +1,1478 @@
...
+  void deallocate(void *Ptr) {}
+
+  /// Add predefined inline asm operand, e.g. %0
+  unsigned addInlineAsmOperand(StringRef Name) {
+    InlineAsmIdentifierInfo &II = AsmBuiltinIdTable.get(Name);
+    InlineAsmOperands.push_back(&II);
+    return InlineAsmOperands.size() - 1;
+  }
+
+  /// Lookup predefined identifiers.
```
Comment (zhimingwang36) [inline clang/lib/DPCT/Asm/AsmParser.h]: Maybe better to add comment on format for Name parameter.

File: clang/lib/DPCT/Asm/AsmParser.h [hunk 21]
```diff
@@ -0,0 +1,1478 @@
...
+  InlineAsmBuiltinType *getF32Type() {
+    return getBuiltinType(InlineAsmBuiltinType::TK_f32);
+  }
+
+  InlineAsmBuiltinType *getF64Type() {
+    return getBuiltinType(InlineAsmBuiltinType::TK_f64);
+  }
+};
+
+/// Introduces a new scope for parsing.
```
Comment (zhimingwang36) [inline clang/lib/DPCT/Asm/AsmParser.h]:
```suggestion
/// Introduces a new scope for parsing when meet ...? (give more detail with example on tree strusture of scope.
```

File: clang/lib/DPCT/Asm/AsmParser.h [hunk 22]
```diff
@@ -0,0 +1,1478 @@
...
+using InlineAsmStmtResult = clang::ActionResult<InlineAsmStmt *>;
+using InlineAsmExprResult = clang::ActionResult<InlineAsmExpr *>;
+
+inline InlineAsmExprResult AsmExprError() { return InlineAsmExprResult(true); }
+inline InlineAsmStmtResult AsmStmtError() { return InlineAsmStmtResult(true); }
+inline InlineAsmTypeResult AsmTypeError() { return InlineAsmTypeResult(true); }
+inline InlineAsmDeclResult AsmDeclError() { return InlineAsmDeclResult(true); }
+
+// clang-format off
+namespace asmprec {
```
Comment (zhimingwang36) [inline clang/lib/DPCT/Asm/AsmParser.h]:
```suggestion
namespace asm_precedence {
```

File: clang/lib/DPCT/Asm/AsmIdentifierTable.h [hunk 23]
```diff
@@ -0,0 +1,193 @@
...
+  StringRef getName() const { return StringRef(getNameStart(), getLength()); }
+
+  /// If this is a source-language token (e.g. 'for'), this API
+  /// can be used to cause the lexer to map identifiers to source-language
+  /// tokens.
+  asmtok::TokenKind getTokenID() const {
+    return static_cast<asmtok::TokenKind>(TokenID);
+  }
+
+  /// Return true if this token is a keyword in asm.
```
Comment (zhimingwang36) [inline clang/lib/DPCT/Asm/AsmIdentifierTable.h]: not all keyword are instruction

File: clang/lib/DPCT/Asm/AsmIdentifierTable.h [hunk 24]
```diff
@@ -0,0 +1,193 @@
...
+
+  InlineAsmIdentifierInfo &get(StringRef Name, asmtok::TokenKind TokenCode) {
+    InlineAsmIdentifierInfo &II = get(Name);
+    II.TokenID = TokenCode;
+    assert(II.TokenID == (unsigned)TokenCode && "TokenCode too large");
+    return II;
+  }
+
+  using iterator = HashTableTy::const_iterator;
+  using const_iterator = HashTableTy::const_iterator;
```
Comment (zhimingwang36) [inline clang/lib/DPCT/Asm/AsmIdentifierTable.h]:
```suggestion

```

File: clang/lib/DPCT/Asm/AsmIdentifierTable.h [hunk 25]
```diff
@@ -0,0 +1,193 @@
...
+  bool contains(StringRef Name) const { return HashTable.contains(Name); }
+
+  /// Populate the identifier table with info about the asm keywords.
+  void AddKeywords();
+};
+
+} // namespace clang::dpct
+
+namespace llvm {
+
```
Comment (zhimingwang36) [inline clang/lib/DPCT/Asm/AsmIdentifierTable.h]: Line 160 ~ 192, consider to remove if not necessary?

File: clang/lib/DPCT/Asm/AsmTokenKinds.def [hunk 26]
```diff
@@ -0,0 +1,337 @@
...
+#undef ALT_FLOAT_TYPE
+#undef PACKED_FLOAT
+#undef PACKED_INTEGER
+#undef PACKED_TYPE
+#undef SUBWORD_TYPE
+#undef BUILTIN_TYPE
+#undef INSTRUCTION
+#undef KEYWORD
+#undef PUNCTUATOR
+#undef TOK
```
Comment (zhimingwang36) [inline clang/lib/DPCT/Asm/AsmTokenKinds.def:344]: there is one more need to undef.  (15)

File: clang/lib/DPCT/Asm/AsmIdentifierTable.cpp [hunk 27]
```diff
@@ -0,0 +1,55 @@
...
+
+void InlineAsmIdentifierTable::AddKeywords() {
+#define KEYWORD(X, Y) (void)get(#X, asmtok::kw_##X);
+#define INSTRUCTION(X) (void)get(#X, asmtok::op_##X);
+#include "AsmTokenKinds.def"
+}
+
+bool InlineAsmIdentifierInfo::isInstruction() const {
+  switch (getTokenID()) {
+#define KEYWORD(X, Y)                                                          \
```
Comment (zhimingwang36) [inline clang/lib/DPCT/Asm/AsmIdentifierTable.cpp]: is Line 32 not need?

File: clang/lib/DPCT/Asm/AsmIdentifierTable.h [hunk 28]
```diff
@@ -0,0 +1,193 @@
...
+  /// Retrieve the external identifier lookup object, if any.
+  InlineAsmIdentifierInfoLookup *getExternalIdentifierLookup() const {
+    return ExternalLookup;
+  }
+
+  llvm::BumpPtrAllocator &getAllocator() { return HashTable.getAllocator(); }
+
+  /// Return the identifier token info for the specified named
+  /// identifier.
```
Comment (zhimingwang36) [inline clang/lib/DPCT/Asm/AsmIdentifierTable.h:114]: Update the comments, 1st try to lookup, if fail, then insert a new record

File: clang/lib/DPCT/Asm/AsmLexer.h [hunk 29]
```diff
@@ -0,0 +1,136 @@
...
+#include "llvm/ADT/StringRef.h"
+#include "llvm/Support/MemoryBufferRef.h"
+#include "llvm/Support/SMLoc.h"
+#include "llvm/Support/raw_ostream.h"
+
+namespace clang::dpct {
+
+/// InlineAsmLexer - This provides a simple interface that truns a text buffer
+/// into a stream of tokens. This provides no support for buffering, or
+/// buffering/seeking of tokens, only forward lexing is supported.
```
Comment (zhimingwang36) [inline clang/lib/DPCT/Asm/AsmLexer.h:29]:
```suggestion
/// buffering, only forward lexing is supported.
```

File: clang/lib/DPCT/Asm/AsmLexer.h [hunk 30]
```diff
@@ -0,0 +1,136 @@
...
+  const InlineAsmToken &peekAhead(unsigned N) {
+    assert(CachedLexPos + N > CachedTokens.size() && "Confused caching.");
+    for (size_t C = CachedLexPos + N - CachedTokens.size(); C > 0; --C) {
+      CachedTokens.push_back(InlineAsmToken());
+      lex(CachedTokens.back());
+    }
+    return CachedTokens.back();
+  }
+
+  const InlineAsmToken &lookAhead(unsigned N) {
```
Comment (zhimingwang36) [inline clang/lib/DPCT/Asm/AsmLexer.h]: if not used, we can delete

File: clang/lib/DPCT/Asm/AsmLexer.h [hunk 31]
```diff
@@ -0,0 +1,136 @@
...
+  /// lex() should be invoked.
+  CachedTokensTy::size_type CachedLexPos = 0;
+
+public:
+  InlineAsmLexer(llvm::MemoryBufferRef Input);
+  InlineAsmLexer(const InlineAsmLexer &) = delete;
+  InlineAsmLexer &operator=(const InlineAsmLexer &) = delete;
+  ~InlineAsmLexer();
+
+  void setBuffer(StringRef Buf);
```
Comment (zhimingwang36) [inline clang/lib/DPCT/Asm/AsmLexer.h]:
```suggestion
  
```

File: clang/lib/DPCT/Asm/AsmLexer.h [hunk 32]
```diff
@@ -0,0 +1,136 @@
...
+  InlineAsmIdentifierInfo *
+  lookupIdentifierInfo(InlineAsmToken &Identifier) const;
+
+private:
+  /// FormTokenWithChars - When we lex a token, we have identified a span
+  /// starting at BufferPtr, going to TokEnd that forms the token.  This method
+  /// takes that range and assigns it to the token as its location and size.  In
+  /// addition, since tokens cannot overlap, this also updates BufferPtr to be
+  /// TokEnd.
+  void formTokenWithChars(InlineAsmToken &Result, const char *TokEnd,
```
Comment (zhimingwang36) [inline clang/lib/DPCT/Asm/AsmLexer.h]:
```suggestion
  void formToken(InlineAsmToken &Result, const char *TokEnd,
```

File: clang/lib/DPCT/Asm/AsmLexer.cpp [hunk 33]
```diff
@@ -0,0 +1,399 @@
...
+
+static inline bool isIdentifierStart(unsigned char C) {
+  return clang::isAsciiIdentifierStart(C, /*AllowDollar*/ true) || C == '%';
+}
+
+static inline bool isIdentifierContinue(unsigned char C) {
+  return clang::isAsciiIdentifierContinue(C, /*AllowDollar*/ true);
+}
+
+void InlineAsmLexer::cleanIdentifier(SmallVectorImpl<char> &Buf,
```
Comment (zhimingwang36) [inline clang/lib/DPCT/Asm/AsmLexer.cpp:41]: processSpecialChars?

File: clang/lib/DPCT/Asm/AsmLexer.cpp [hunk 34]
```diff
@@ -0,0 +1,399 @@
...
+
+bool InlineAsmLexer::isHexLiteral(const char *Start) const {
+  char C = getChar(Start);
+  if (C != 0)
+    return false;
+  C = getChar(Start + 1);
+  return (C == 'x' || C == 'X');
+}
+
+bool InlineAsmLexer::lex(InlineAsmToken &Result) {
```
Comment (zhimingwang36) [inline clang/lib/DPCT/Asm/AsmLexer.cpp:108]: if look ahead is not required, need clean up the lex() function.

File: clang/lib/DPCT/Asm/AsmLexer.cpp [hunk 35]
```diff
@@ -0,0 +1,399 @@
...
+    II = getIdentifierInfo(IdentifierBuffer);
+  }
+  Identifier.setIdentifier(II);
+  Identifier.setKind(II->getTokenID());
+  return II;
+}
+
+bool InlineAsmLexer::isHexLiteral(const char *Start) const {
+  char C = getChar(Start);
+  if (C != 0)
```
Comment (ziranzha) [inline clang/lib/DPCT/Asm/AsmLexer.cpp:78]: Should be '0'

File: clang/lib/DPCT/Asm/AsmLexer.cpp (hunk 33 above)
Comment (ziranzha) [inline clang/lib/DPCT/Asm/AsmLexer.cpp:50]: Should consider the cleaning rule to avoid redefinition.

File: clang/lib/DPCT/Asm/AsmLexer.cpp (hunk 35 above)
Comment (ziranzha) [inline clang/lib/DPCT/Asm/AsmLexer.cpp]:
```suggestion
  if (C != '0')
```

File: clang/lib/DPCT/Asm/AsmLexer.h [hunk 36]
```diff
@@ -0,0 +1,136 @@
...
+    Result.setLength(TokLen);
+    Result.setLocation(SMLoc::getFromPointer(BufferPtr));
+    Result.setKind(Kind);
+    BufferPtr = TokEnd;
+  }
+
+  bool isHexLiteral(const char *Start) const;
+
+  char getChar(const char *Ptr) const {
+    if (Ptr == BufferEnd)
```
Comment (ziranzha) [inline clang/lib/DPCT/Asm/AsmLexer.h]:
```suggestion
    if (Ptr >= BufferEnd)
```

File: clang/lib/DPCT/Asm/AsmParser.cpp [hunk 37]
```diff
@@ -0,0 +1,1277 @@
...
+
+  if (ExpectAndConsume(asmtok::comma))
+    return AsmStmtError();
+
+  SmallVector<InlineAsmExpr *, 4> InputOperands;
+
+  while (true) {
+    InlineAsmExprResult Operand = ParseExpression();
+    if (Operand.isInvalid())
+      return true;
```
Comment (ziranzha) [inline clang/lib/DPCT/Asm/AsmParser.cpp]:
```suggestion
      return AsmStmtError();
```

File: clang/lib/DPCT/Asm/AsmParser.cpp [hunk 38]
```diff
@@ -0,0 +1,1277 @@
...
+    if (PredOutput.isInvalid())
+      return AsmStmtError();
+  }
+
+  if (ExpectAndConsume(asmtok::comma))
+    return AsmStmtError();
+
+  SmallVector<InlineAsmExpr *, 4> InputOperands;
+
+  while (true) {
```
Comment (ziranzha) [inline clang/lib/DPCT/Asm/AsmParser.cpp]:
```suggestion
  while (ExpectAndConsume(asmtok::comma)) {
```

File: clang/lib/DPCT/Asm/AsmParser.cpp [hunk 39]
```diff
@@ -0,0 +1,1277 @@
...
+InlineAsmExprResult InlineAsmParser::ParseCastExpression() {
+  InlineAsmExprResult Res;
+  auto SavedKind = Tok.getKind();
+  switch (SavedKind) {
+  case asmtok::l_paren:
+    ConsumeParen();
+    if (Tok.isOneOf(asmtok::kw_s64, asmtok::kw_u64)) {
+      InlineAsmBuiltinType *CastTy =
+          Tok.is(asmtok::kw_s64) ? Context.getS64Type() : Context.getU64Type();
+      ConsumeParen();
```
Comment (ziranzha) [inline clang/lib/DPCT/Asm/AsmParser.cpp]:
```suggestion
      if (!ExpectAndConsume(r_paren))
        return AsmExprError();
```

**Review by zhimingwang36 — APPROVED** [review]
LGTM

File: clang/lib/DPCT/Asm/AsmParser.h [hunk 40]
```diff
@@ -0,0 +1,1379 @@
...
+  /// ExitScope - Pop a scope off the scope stack.
+  void ExitScope() {
+    assert(getCurScope());
+    InlineAsmScope *OldScope = getCurScope();
+    if (OldScope) {
+      CurScope = OldScope->getParent();
+      if (NumCachedScopes == ScopeCacheSize)
+        delete OldScope;
+      else
+        ScopeCache[NumCachedScopes++] = OldScope;
```
Comment (ziranzha) [inline clang/lib/DPCT/Asm/AsmParser.h]: Need explicitly call the destructor.

**Review by ziranzha — APPROVED** [review]

**Pull request merged**